src/x/serialize/encoder.go

```diff
@@ -80,53 +84,88 @@ func newTagEncoder(
 ) TagEncoder {
 	b := make([]byte, 0, opts.InitialCapacity())
 	cb := newFn(nil, nil)
-	return &encoder{
-		buf:          bytes.NewBuffer(b),
-		checkedBytes: cb,
-		opts:         opts,
-		pool:         pool,
+	e := &encoder{
+		buf:                 b,
+		checkedBytes:        cb,
+		maxTagLiteralLength: int(opts.TagSerializationLimits().MaxTagLiteralLength()),
+		maxNumberTags:       int(opts.TagSerializationLimits().MaxNumberTags()),
+		pool:                pool,
 	}
+	e.bufUInt16 = e.bufUInt16Const[:]
+	return e
 }
 
-func (e *encoder) Encode(srcTags ident.TagIterator) error {
-	if e.checkedBytes.NumRef() > 0 {
-		return errTagEncoderInUse
+func (e *encoder) Encode(tagsIter ident.TagIterator) error {
+	numTags := tagsIter.Remaining()
+	if err := e.writeHeader(numTags); err != nil {
+		return err
 	}
 
-	tags := srcTags.Duplicate()
-	defer tags.Close()
+	if sliceIter, ok := tagsIter.(interface {
+		SliceIter() []ident.Tag
```

Comment:
Does the commit log write path have a SliceIter that backs it?  Or Is it only the tests that benefit from this speed up?

Reply:
Yeah it does I think unless I’m misunderstanding something: https://github.com/m3db/m3/blob/master/src/dbnode/persist/fs/commitlog/writer.go#L137

Reply:
I don't see this on that type though? e.g. searching for `SliceIter() []ident.Tag` in https://github.com/m3db/m3/blob/master/src/x/ident/tag_iterator.go yields no results

Reply:
I see, should we perhaps just make this another method, i.e. `EncodeSlice(...)` or similar?

Reply:
I think the interface assertion is the cleanest personally (the assertion is very cheap since it only needs to check if one method exists) but we can add a new method if you feel strongly